airbyte-integrations/connectors/source-orb/source_orb/source.py

```diff
@@ -177,6 +180,257 @@ def transform_record(self, subscription_record):
+        chunk_start_date = chunk_end_date
+
+
+class SubscriptionUsage(IncrementalOrbStream):
+    """
+    API Docs: https://docs.withorb.com/docs/orb-docs/api-reference/operations/get-a-subscription-usage
+    """
+
+    cursor_field = "timeframe_start"
+
+    def __init__(
+        self,
+        subscription_usage_grouping_key: Optional[str] = None,
+        plan_id: Optional[str] = None,
+        end_date: Optional[pendulum.DateTime] = None,
+        **kwargs,
+    ):
+
+        self.subscription_usage_grouping_key = subscription_usage_grouping_key
+        self.plan_id = plan_id
+        self.end_date = end_date
+        super().__init__(**kwargs)
+
+    @property
+    def primary_key(self) -> Iterable[str]:
+        key = ["subscription_id", "billable_metric_id", "timeframe_start"]
+
+        # If a grouping key is present, it should be included in the primary key
+        if self.subscription_usage_grouping_key:
+            key.append(self.subscription_usage_grouping_key)
+
+        return key
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        # use a regex capture group to extract the subscription ID from the request URL
+        # so that we can include it in the returned record
+        subscription_id = re.search("/subscriptions/(.*)/usage", response.request.url).group(1)
+        self.logger.debug("parsed subscription id from url: %s", subscription_id)
+
+        # Records are in a container array called data
+        response_json = response.json()
+        records = response_json.get("data", [])
+        for record in records:
+            # for each top level record, there can be multiple sub-records depending
+            # on granularity and other input params. Oftentimes there will just
+            # be a single subrecord, but this approach creates one resulting
+            # record for each regardless of how many subrecords exist.
+            subrecords = record.get("usage", [])
+            del record["usage"]
+            for subrecord in subrecords:
+                # skip records that don't contain any actual usage
+                if subrecord.get("quantity", 0) > 0:
+                    yield self.transform_record(record, subrecord, subscription_id)
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        # This API endpoint is not paginated, so there will never be a next page
+        return None
+
+    def transform_record(self, parent_record, sub_record, subscription_id):
+        self.logger.debug("incoming subscription usage record: %s", parent_record)
+
+        # Merge the parent record with the sub record
+        record = {**parent_record, **sub_record}
+
+        # Add the subscription ID to the record
+        record["subscription_id"] = subscription_id
+
+        # Un-nest billable_metric -> name,id into billable_metric_name and billable_metric_id
```

Comment:
Why do you un-nest the billable metric in the source instead of letting the normalization handle it?

Reply:
I wanted to use billable_metric_id as part of the primary key, and felt like it would be clearer to un-nest everything at the record level so that the primary key path and record schema was easy to work with. Is there a good reason or best practice for keeping it nested that I should try to follow instead?

Reply:
nope, I asked out of curiosity. We leave that decision to the connector developer. This is totally fine